### PR TITLE
Enables basic generation of Unions

### PR DESCRIPTION
**Pull request by Luke-Nukem**
Enables basic generation of Unions for Rust. It is feature gated so needs a compile-time option (until unions are stable in Rust).

Also updates Readme by rewording and clarifying some sections, and adds section about Unions.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,16 +1,26 @@
...
+The `*.toml` is what is used to pass various settings and options to gir for use when generating the bindings - you will likely need to write one to suit your needs, for an example you can take a look to [gtk-rs/sys/gir-gtk.toml](https://github.com/gtk-rs/sys/blob/master/conf/gir-gtk.toml).
 
-In FFI mode, Gir generates as much as it can. So in this mode, the TOML file is mostly used to ignore some objects. To do so, you need to add its fullname to an `ignore` array. Example:
+## `gir` Modes
+
+There are two main modes of generation for `gir`; *FFI* and *API*.
+
+The *FFI* mode is what creates the low-level FFI bindings from the supplied `*.gir` file - these are essentially direct calls in to the related C library and are typically unsafe. The resulting crate is typically appended with `-sys`.
+
+The *API* mode generates another crate for a layer on top of these unsafe (*sys*) bindings whihc makes them safe for use in general Rust.
```
Comment (EPashkin) [inline README.md]: which

File: src/codegen/sys/lib_.rs [hunk 2]
```diff
@@ -317,19 +317,46 @@ fn generate_unions(w: &mut Write, env: &Env, items: &[&Union]) -> Result<()> {
...
+                        name = c_type
+                    ));
+
+                    for line in lines {
+                        try!(writeln!(w, "{}{}", comment, line));
+                    }
+                    try!(writeln!(w, "{}}}\n", comment));
+                }
+            }
+            if !cfg!(feature = "use_unions") {
```
Comment (EPashkin) [inline src/codegen/sys/lib_.rs]: else ?

**Comment by EPashkin** [comment]
In glib generated `GValue` changed in bad way (pointers 32bit on 32bit targets).
```diff
 #[repr(C)]
 pub struct GValue {
     pub g_type: GType,
-    pub data: [u64; 2],
+    pub data: [gpointer; 2],
}
```

**Comment by EPashkin** [comment]
and in gobject-sys too

**Comment by EPashkin** [comment]
Sorry, hit wrong button

**Comment by EPashkin** [comment]
Also better if it can falback to old mode ("_truncated_record_marker") if one element types has `Err` in type
```diff
-#[repr(C)]
-pub struct GVariantBuilder {
-    _truncated_record_marker: c_void,
-    //union,
-}
+//#[repr(C)]
+//pub struct GVariantBuilder {
+//    pub : [TypeId { ns_id: 0, id: 72 } *.],
+//}
 ```

**Comment by EPashkin** [comment]
.. while list of union fields also better stays on fallback (it simplified finding source of problem).

**Comment by Luke-Nukem** [comment]
@EPashkin I'm working on solving some of these issues in a new branch that does a more complete union generation. Will cancel this pull soon and make a new one.

Can you explain where

```
 #[repr(C)]
 pub struct GValue {
     pub g_type: GType,
-    pub data: [u64; 2],
+    pub data: [gpointer; 2],
}
```

Is coming from? I don't seem to be getting it generated using my branch so i can't verify it is solved. (it is the GLib gir, yes?)

**Comment by Luke-Nukem** [comment]
Closing pull request in favour of new branch with full Union generation and feature-gating.

**Comment by Luke-Nukem** [comment]
Closing pull request in favour of new branch with full Union generation and feature-gating.

**Pull request closed without merging**